draft-ietf-quic-http.md

```diff
@@ -430,22 +473,87 @@ The flags defined are:
+    stream when the PUSH_PRIORITIZED flag is clear, or a Push ID when the PUSH
+    flag is set.
 
   Stream Dependency:
-  : A 32-bit stream identifier for the request stream that this stream depends
-    on (see {{priority}} and {{!RFC7540}}, Section 5.3).
+  : A 32-bit stream identifier for a dependent request.  This contains the
+    stream ID of a request stream when the PUSH_DEPENDENT flag is clear, or a
+    Push ID when the PUSH_DEPENDENT flag is set.  A request stream ID of 0
+    indicates a dependency on the root stream. For details of dependencies,
+    see {{priority}} and {!RFC7540}}, Section 5.3).
 
   Weight:
   : An unsigned 8-bit integer representing a priority weight for the stream (see
     {{!RFC7540}}, Section 5.3). Add one to the value to obtain a weight between
     1 and 256.
 
-A PRIORITY frame MUST have a payload length of nine octets.  A PRIORITY frame
-of any other length MUST be treated as a connection error of type
-HTTP_MALFORMED_PRIORITY.
+A PRIORITY frame identifies a request to priotize, and a request upon which that
+request is dependent.  A request is identified by the stream ID of a request
+when the corresponding PUSH_PRIORITIZED or PUSH_DEPENDENT flag is not set.
+Setting the PUSH or PUSH_DEPENDENT flag causes the frame to identify a
+PUSH_PROMISE using a Push ID (see {{frame-push-promise}} for details).
+
+A PRIORITY frame MAY identify no request by using a stream ID of 0; as in
+{{!RFC7540}}, this makes the request dependent on the root of the dependency
+tree.
+
+A PRIORITY frame that does not reference a request MUST be treated as a
+HTTP_MALFORMED_PRIORITY error, unless it references stream ID 0.  A PRIORITY
+that sets a PUSH_PRIORITIZED or PUSH_DEPENDENT flag, but then references a
+non-existent Push ID MUST be treated as a HTTP_MALFORMED_PRIORITY error.
+
+The length of a PRIORITY frame is 9 octets.  A PRIORITY frame with any other
+length MUST be treated as a connection error of type HTTP_MALFORMED_PRIORITY.
+
+
+### CANCEL_REQUEST {#frame-cancel-request}
+
+The CANCEL_REQUEST frame (type=0x3) is used to request cancellation of a request
+prior to the response stream being created.  This frame can be used to cancel
+server push requests that are initiated with a PUSH_PROMISE.
+
+The CANCEL_REQUEST frame identifies a server push request by Push ID (see
+{{frame-push-promise}}).
+
+When a server receives this frame, it aborts sending the response for the
+identified server push.  If the server has not yet started to send the server
+push, it can use the receipt of a CANCEL_REQUEST frame to avoid opening a
+stream.  If the push stream has been opened by the server, the server SHOULD
+sent a QUIC RST_STREAM frame on those streams and cease transmission of the
+response.
+
+A server can send this frame to indicate that it won't be sending a response
+prior to creation of a push stream.  Once the push stream has been created, a
+RST_STREAM with a CANCELLED code can be used instead.
```

Comment:
I've gone with SHOULD.  There isn't any point in MUST here, because it's not a correctness issue, it's just pointless to use CANCEL_PUSH at this point.